SpatiaLite support for importing layers with binary fields

audun opened the pull request:
## Description

Adds support for binary fields to SpatiaLite like in PostGIS. Fixes #36705.
prepare_commit.sh has been run.

A few thoughts about the changes:

~Since there are currently no unit tests (that I am aware of) for Spatialite~ There are plenty of unit tests run during the Python tests, but since I wasn't aware of them, I wanted to make the changes minimal and easy to verify.  There should be absolutely no functional changes in anything except for supporting ByteArray fields. I have tested all the changed code paths manually, including the error paths (by intentionally introducing errors in the SQL).

In the UPDATE statement I use sqlite3_bind_blob instead of using a binary literal. It's possible in SQLite to write blob literals by first encoding it as hexadecimal, but I think it's more difficult to verify the correctness and would have much worse performance than simply binding the value directly.

By taking that approach further and using bind instead of string literals in the entire statement I believe it's possible to move the binding of values in both the INSERT and the UPDATE statement to a common function. Should result in less code and better performance for UPDATE, but doing that would cause changes to the current functionality.

Also as @m-kuhn mentioned, using RAII would simplify the necessity of using finalize properly everywhere.

That must be in another PR or three though.

<!--
  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

audun committed:
SpatiaLite support for importing layers with binary fields

Fixes #36705

nyalldawson commented on src/providers/spatialite/qgsspatialiteprovider.cpp:
```diff
@@ -4564,9 +4591,61 @@ bool QgsSpatiaLiteProvider::changeAttributeValues( const QgsChangedAttributesMap
+          // This will only happen if the above code is changed to bind more types,
+          // but the programmer has forgotten to handle the type here. Fatal error.
+          sqlite3_finalize( stmt );
+          throw std::runtime_error( "Attempting to bind unhandled type." );
```
Better to Q_ASSERT( false ) here

audun replied:
It seems like the Q_ASSERT is not checked in release mode, so if the programmer forgets to handle the type correctly here it could then silently cause user data loss. Since this error can never occur because of user input or data I think it's better to visibly let it bubble up to the "Caught unhandled std::exception" handler.

What do you think?

nyalldawson reviewed (commented):
Can you add some new unit tests covering this functionality?

audun commented:
Yes, absolutely. I was wondering how to approach writing tests, but incorporating some new ones in Python should be easy. This test suite will give me confidence enough in refactoring out the duplicate code eventually too. Thanks for reviewing the PR so far.

I have a question: It appears that the test for the Spatialite provider, `tests/src/python/test_provider_spatialite.py` is not run during the build. Seems like that might be the case for all the DB providers, but perhaps Spatialite should be an exception as it does not need a server?

Locally the test fails in an unrelated test, when it tries to remove the database file in testPKNotInt test. It seems like that test does the cleanup differently than all the rest, for some reason. Commenting out that `shutil.rmtree(basepath)` call makes the entire test suite pass. I'll commit a fix for that too.

audun committed:
Clean up temp directory in test tear down

testPKNotInt was failing on Windows because the file was still open

nyalldawson commented:
> I have a question: It appears that the test for the Spatialite provider, tests/src/python/test_provider_spatialite.py is not run during the build.

It is run on Travis CI. None of the tests are run during the build, you need to manually run all of them post-build. See https://docs.qgis.org/testing/en/docs/developers_guide/unittesting.html

elpaso commented on src/providers/spatialite/qgsspatialiteprovider.cpp:
```diff
@@ -44,6 +44,8 @@ email                : a.furieri@lqt.it
 #include <QDir>
 #include <QRegularExpression>
 
+#include <stdexcept>
```
not needed anymore?

audun replied:
Depends on the resolution of the other comment, but I have followed the advice there in the current version so it's gone at the moment.

audun commented:
I have now added test cases exercising the binary layers in the same manner that is done in TestPyQgsPostgresProvider (exposing a bug 🐛) and rebased it on latest master.

audun commented:
> > I have a question: It appears that the test for the Spatialite provider, tests/src/python/test_provider_spatialite.py is not run during the build.
> 
> It is run on Travis CI. None of the tests are run during the build, you need to manually run all of them post-build. See https://docs.qgis.org/testing/en/docs/developers_guide/unittesting.html

Sorry for being unclear. I meant on the build server during CI, but I was looking only on the Azure build and not on Travis. 
The only test I can see on Azure with Spatialite in the name is [PyQgsDBManagerSpatialite](https://dev.azure.com/qgis/QGIS/_build/results?buildId=5821&view=logs&j=c55c25a0-2401-5b7f-f792-650180689337&t=e313f7aa-4452-5f33-1018-5aea7c0f02b7&l=1069).  I expected to see PyQgsSpatialiteProvider and PyQgsProviderConnectionSpatialite too, like when I am running tests locally.

Anyway, on Travis it is there as expected:

`        Start 524: PyQgsSpatialiteProvider
524/644 Test #524: PyQgsSpatialiteProvider ...........................   Passed    4.26 sec`

m-kuhn commented:
The azure ones were integrated only recently and we just enabled to avoid regressions. This one was probably failing but it was not investigated why.
It would be great to enable more tests on Azure. If you want to give it a go, remove it from [this list](https://github.com/qgis/QGIS/blob/master/.ci/azure-pipelines/azure-pipelines.yml#L4) and see what's required to stabilize it.

audun commented:
Perhaps it had the same problem that I had to fix to get it running locally on Windows? I can try to remove it later, but currently it seems like there is a very long queue of builds so might not the best time to experiment.

nirvn reviewed (approved):
Looking good. +1 to merge when travis is green.

audun commented:
Perfect, thanks! 

Tested to use the conflict resolution tool in GitHub, but it created a merge commit. Did a rebase onto latest master to remove the merge commit as recommended in the contributor guide.

m-kuhn commented on src/providers/spatialite/qgsspatialiteprovider.cpp:
```diff
@@ -4623,9 +4657,61 @@ bool QgsSpatiaLiteProvider::changeAttributeValues( const QgsChangedAttributesMap
+    }
+
+    // bind variables not handled directly in the string
+    QMap<int, QVariant>::iterator i;
```
Can you use a const iterator here?

audun replied:
Good idea. I see auto used above, so that makes it even easier.

m-kuhn commented on src/providers/spatialite/qgsspatialitefeatureiterator.cpp:
```diff
@@ -588,6 +588,14 @@ QVariant QgsSpatiaLiteFeatureIterator::getFeatureAttribute( sqlite3_stmt *stmt,
+    // BLOB value
+    int blob_size = sqlite3_column_bytes( stmt, ic );
+    const void *blob = sqlite3_column_blob( stmt, ic );
+    return QByteArray( ( const char * ) blob, blob_size );
```
```suggestion
    return QByteArray( static_cast<const char *>( blob ), blob_size );
```

Super minor nitpicking, static_casts are the preferred way to handle this. (also in qgsspatialiteprovider.cpp )

audun replied:
I totally agree, I actually wrote it like that originally, but I try to follow the style close to the code, and a few lines down there's another C-style cast.

```
    // TEXT value
    const QString txt = QString::fromUtf8( ( const char * ) sqlite3_column_text( stmt, ic ) );
```

I'll follow your suggestion here, and we can perhaps make a separate PR making it consistently C++-style casts elsewhere.

nyalldawson commented:
Superseded by https://github.com/qgis/QGIS/pull/36876
